File: backend_py/src/edu_backend/services/docx_parser.py
"""
DOCX educational-design parser — Python port of web/src/utils/docxParser.ts

Extracts structured data (title, total hours, modules with ESCO skills) from a
.docx educational design document. Mirrors the regex-based logic of the
frontend mammoth-based parser so that API consumers get identical results.

Keep this file in sync with web/src/utils/docxParser.ts when the document
format changes.
"""
from __future__ import annotations

import io
import re
from dataclasses import dataclass, field
from typing import Literal

from docx import Document


SkillType = Literal["essential", "optional"]


@dataclass
class ESCOSkill:
    code: str
    name: str
    type: SkillType


@dataclass
class Module:
    number: int
    title: str
    hours: int
    content: str = ""
    activities: str = ""
    skills: list[ESCOSkill] = field(default_factory=list)


@dataclass
class ParsedDocument:
    documentTitle: str
    totalHours: int
    modules: list[Module]


def parse_docx_bytes(data: bytes) -> ParsedDocument:
    """Parse a .docx file (bytes) into a ParsedDocument."""
    doc = Document(io.BytesIO(data))
    text = "\n".join(p.text for p in doc.paragraphs)
    return parse_educational_design(text)


def parse_educational_design(text: str) -> ParsedDocument:
    lines = [ln for ln in text.split("\n") if ln.strip()]

    title = ""
    occupation_match = re.search(
        r"(?:για\s+)?(στελέχη[^,\n.]+|Ειδικός[^,\n.]+)", text, re.IGNORECASE
    )
    if occupation_match:
        occupation = occupation_match.group(1).strip()
        title = f"Πρόγραμμα επαγγελματικής κατάρτισης για {occupation}"
    else:
        for line in lines:
            if "Ειδικός" in line or "Πρόγραμμα" in line:
                short_match = re.search(
                    r"(Ειδικός[^,\n]+|Πρόγραμμα[^για]*για[^,\n]+)",
                    line,
                    re.IGNORECASE,
                )
                title = short_match.group(1).strip() if short_match else line.strip()[:80]
                break

    if not title and lines:
        title = lines[0].strip()[:80]

    total_hours = 0
    hours_match = re.search(r"Συνολικές\s*[Ωώ]ρες[:\s]*(\d+)", text, re.IGNORECASE)
    if hours_match:
        total_hours = int(hours_match.group(1))

    modules = _extract_modules(text)

    if not total_hours:
        total_hours = sum(m.hours for m in modules)

    return ParsedDocument(documentTitle=title, totalHours=total_hours, modules=modules)


def _extract_modules(text: str) -> list[Module]:
    thematic_start = text.find("Θεματικές Ενότητες")
    working_text = text[thematic_start:] if thematic_start >= 0 else text

    module_pattern = re.compile(
        r"(\d+)\.\s+([^\n]+)\n[^\d]*?(\d+)\s*ώρες", re.DOTALL
    )
    matches = list(module_pattern.finditer(working_text))

    if not matches:
        return _extract_modules_alternative(text)

    modules: list[Module] = []
    for i, match in enumerate(matches):
        number = int(match.group(1))
        m_title = match.group(2).strip()
        hours = int(match.group(3))

        start_idx = match.end()
        end_idx = matches[i + 1].start() if i + 1 < len(matches) else len(working_text)
        section_text = working_text[start_idx:end_idx]

        content_match = re.search(
            r"Περιεχόμενο:\s*([^🎮Δ]+?)(?=Δραστηριότητες|ESCO|$)",
            section_text,
            re.DOTALL,
        )
        content = content_match.group(1).strip() if content_match else ""

        activities_match = re.search(
            r"Δραστηριότητες:\s*([^🎯E]+?)(?=ESCO|$)",
            section_text,
            re.DOTALL,
        )
        activities = activities_match.group(1).strip() if activities_match else ""

        skills_match = re.search(
            r"ESCO\s*Skills[^:]*:\s*([\s\S]*?)(?=\d+\.|$)",
            section_text,
        )
        skills_text = skills_match.group(1) if skills_match else ""
        skills = _parse_skills(skills_text)

        modules.append(
            Module(
                number=number,
                title=m_title,
                hours=hours,
                content=content,
                activities=activities,
                skills=skills,
            )
        )

    modules.sort(key=lambda m: m.number)
    return modules


def _extract_modules_alternative(text: str) -> list[Module]:
    modules: list[Module] = []
    lines = text.split("\n")
    current: Module | None = None
    content_buffer = ""

    module_header = re.compile(r"^(\d+)\.\s+(.+?)(?:\s*[-–]\s*(\d+)\s*ώρες)?$")
    next_line_hours = re.compile(r"^\s*(\d+)\s*ώρες", re.IGNORECASE)
    inline_hours = re.compile(r"(\d+)\s*ώρες", re.IGNORECASE)

    for i, raw_line in enumerate(lines):
        line = raw_line.strip()
        next_line = lines[i + 1].strip() if i + 1 < len(lines) else ""

        m = module_header.match(line)
        if m:
            has_hours_in_line = m.group(3) is not None
            next_match = next_line_hours.match(next_line)
            has_hours_in_next_line = next_match is not None
            lower = line.lower()
            has_module_keyword = (
                "ενότητα" in lower or "θεματική" in lower or "κεφάλαιο" in lower
            )

            if not has_hours_in_line and not has_hours_in_next_line and not has_module_keyword:
                if current is not None:
                    content_buffer += line + "\n"
                continue

            if current is not None and current.number and current.title:
                current.content = content_buffer.strip()
                modules.append(current)

            if has_hours_in_line:
                hours = int(m.group(3))
            elif next_match:
                hours = int(next_match.group(1))
            else:
                hours = 0

            current = Module(
                number=int(m.group(1)),
                title=m.group(2).strip(),
                hours=hours,
            )
            content_buffer = ""
            continue

        if current is not None:
            ih = inline_hours.search(line)
            if ih and not current.hours:
                current.hours = int(ih.group(1))

            if current is not None and ("ESCO" in line or "[E]" in line or "[O]" in line):
                current.skills.extend(_parse_skills(line))

            content_buffer += line + "\n"

    if current is not None and current.number and current.title:
        current.content = content_buffer.strip()
        modules.append(current)

    return modules


def _parse_skills(text: str) -> list[ESCOSkill]:
    skills: list[ESCOSkill] = []

    skill_pattern = re.compile(r"([^\[\]\n,]+)\s*\[([EO])\]")
    for match in skill_pattern.finditer(text):
        name = match.group(1).strip()
        skill_type: SkillType = "essential" if match.group(2) == "E" else "optional"
        if name and len(name) > 2:
            skills.append(ESCOSkill(code=_generate_skill_code(name), name=name, type=skill_type))

    if not skills:
        for part in re.split(r"[,\n]", text):
            name = part.strip()
            if name and len(name) > 3 and not re.match(r"^\d", name) and "ESCO" not in name:
                skills.append(
                    ESCOSkill(code=_generate_skill_code(name), name=name, type="essential")
                )

    return skills


def _generate_skill_code(name: str) -> str:
    cleaned = re.sub(r"[^\w\sα-ωά-ώ]", "", name.lower(), flags=re.UNICODE)
    parts = re.split(r"\s+", cleaned)
    code = "-".join(p[:3] for p in parts if p)
    return code[:20]


def to_dict(parsed: ParsedDocument) -> dict:
    """Serialize ParsedDocument to a JSON-friendly dict matching the frontend shape."""
    return {
        "documentTitle": parsed.documentTitle,
        "totalHours": parsed.totalHours,
        "modules": [
            {
                "number": m.number,
                "title": m.title,
                "hours": m.hours,
                "content": m.content,
                "activities": m.activities,
                "skills": [
                    {"code": s.code, "name": s.name, "type": s.type} for s in m.skills
                ],
            }
            for m in parsed.modules
        ],
    }

File: web/src/utils/docxParser.ts
import mammoth from 'mammoth';
import type { Module, ESCOSkill } from '../store/useStore';

interface ParsedDocument {
  title: string;
  totalHours: number;
  modules: Module[];
}

export async function parseDocx(file: File): Promise<ParsedDocument> {
  const arrayBuffer = await file.arrayBuffer();
  const result = await mammoth.extractRawText({ arrayBuffer });
  const text = result.value;

  return parseEducationalDesign(text);
}

function parseEducationalDesign(text: string): ParsedDocument {
  const lines = text.split('\n').filter(line => line.trim());

  // Extract occupation title - look for "στελέχη" or "Ειδικός" pattern
  let title = '';

  // Try to find occupation pattern like "στελέχη διοικητικής υποστήριξης"
  const occupationMatch = text.match(/(?:για\s+)?(στελέχη[^,\n.]+|Ειδικός[^,\n.]+)/i);
  if (occupationMatch) {
    const occupation = occupationMatch[1].trim();
    title = `Πρόγραμμα επαγγελματικής κατάρτισης για ${occupation}`;
  } else {
    // Fallback: look for first line with "Ειδικός" or "Πρόγραμμα"
    for (const line of lines) {
      if (line.includes('Ειδικός') || line.includes('Πρόγραμμα')) {
        // Keep only the short version
        const shortMatch = line.match(/(Ειδικός[^,\n]+|Πρόγραμμα[^για]*για[^,\n]+)/i);
        title = shortMatch ? shortMatch[1].trim() : line.trim().substring(0, 80);
        break;
      }
    }
  }

  if (!title && lines.length > 0) {
    title = lines[0].trim().substring(0, 80);
  }

  // Extract total hours
  let totalHours = 0;
  const hoursMatch = text.match(/Συνολικές\s*[Ωώ]ρες[:\s]*(\d+)/i);
  if (hoursMatch) {
    totalHours = parseInt(hoursMatch[1], 10);
  }

  // Extract modules
  const modules = extractModules(text);

  return {
    title,
    totalHours: totalHours || modules.reduce((sum, m) => sum + m.hours, 0),
    modules,
  };
}

function extractModules(text: string): Module[] {
  const modules: Module[] = [];

  // Find "Θεματικές Ενότητες" section and work from there
  const thematicStart = text.indexOf('Θεματικές Ενότητες');
  const workingText = thematicStart >= 0 ? text.substring(thematicStart) : text;

  // Split by module pattern: "NUMBER. TITLE" followed by hours on next line
  // Pattern matches: "1. Title\n⏱️ 15 ώρες" or "1. Title\n15 ώρες"
  const modulePattern = /(\d+)\.\s+([^\n]+)\n[^\d]*?(\d+)\s*ώρες/g;
  const matches = [...workingText.matchAll(modulePattern)];

  if (matches.length === 0) {
    return extractModulesAlternative(text);
  }

  for (let i = 0; i < matches.length; i++) {
    const match = matches[i];
    const number = parseInt(match[1], 10);
    const title = match[2].trim();
    const hours = parseInt(match[3], 10);

    // Get section text (from this match to next match or end)
    const startIdx = match.index! + match[0].length;
    const endIdx = matches[i + 1]?.index ?? workingText.length;
    const sectionText = workingText.substring(startIdx, endIdx);

    // Extract content
    const contentMatch = sectionText.match(/Περιεχόμενο:\s*([^🎮Δ]+?)(?=Δραστηριότητες|ESCO|$)/s);
    const content = contentMatch ? contentMatch[1].trim() : '';

    // Extract activities
    const activitiesMatch = sectionText.match(/Δραστηριότητες:\s*([^🎯E]+?)(?=ESCO|$)/s);
    const activities = activitiesMatch ? activitiesMatch[1].trim() : '';

    // Extract ESCO skills
    const skillsMatch = sectionText.match(/ESCO\s*Skills[^:]*:\s*([\s\S]*?)(?=\d+\.|$)/);
    const skillsText = skillsMatch ? skillsMatch[1] : '';
    const skills = parseSkills(skillsText);

    modules.push({
      number,
      title,
      hours,
      content,
      activities,
      skills,
    });
  }

  return modules.sort((a, b) => a.number - b.number);
}

function extractModulesAlternative(text: string): Module[] {
  const modules: Module[] = [];

  // Look for numbered sections with hours mentioned
  const lines = text.split('\n');
  let currentModule: Partial<Module> | null = null;
  let contentBuffer = '';

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    const nextLine = lines[i + 1]?.trim() || '';

    // Check for module header patterns - MUST have hours mentioned on same or next line
    // Pattern: "1. Title - 8 ώρες" or "1. Title" followed by "8 ώρες" on next line
    const moduleMatch = line.match(/^(\d+)\.\s+(.+?)(?:\s*[-–]\s*(\d+)\s*ώρες)?$/);

    if (moduleMatch) {
      const hasHoursInLine = moduleMatch[3] !== undefined;
      const hasHoursInNextLine = nextLine.match(/^\s*(\d+)\s*ώρες/i);
      const hasModuleKeyword = line.toLowerCase().includes('ενότητα') ||
                               line.toLowerCase().includes('θεματική') ||
                               line.toLowerCase().includes('κεφάλαιο');

      // Only consider it a module if it has hours OR module keywords
      if (!hasHoursInLine && !hasHoursInNextLine && !hasModuleKeyword) {
        // This is likely a learning objective, not a module - skip
        if (currentModule) {
          contentBuffer += line + '\n';
        }
        continue;
      }

      // Save previous module
      if (currentModule && currentModule.number && currentModule.title) {
        currentModule.content = contentBuffer.trim();
        modules.push(currentModule as Module);
      }

      const hours = hasHoursInLine
        ? parseInt(moduleMatch[3], 10)
        : (hasHoursInNextLine ? parseInt(hasHoursInNextLine[1], 10) : 0);

      currentModule = {
        number: parseInt(moduleMatch[1], 10),
        title: moduleMatch[2].trim(),
        hours,
        content: '',
        activities: '',
        skills: [],
      };
      contentBuffer = '';
      continue;
    }

    // Check for hours pattern
    if (currentModule && line.match(/(\d+)\s*ώρες/i)) {
      const hoursMatch = line.match(/(\d+)\s*ώρες/i);
      if (hoursMatch && !currentModule.hours) {
        currentModule.hours = parseInt(hoursMatch[1], 10);
      }
    }

    // Check for ESCO skills
    if (currentModule && (line.includes('ESCO') || line.includes('[E]') || line.includes('[O]'))) {
      const skills = parseSkills(line);
      currentModule.skills = [...(currentModule.skills || []), ...skills];
    }

    // Accumulate content
    if (currentModule) {
      contentBuffer += line + '\n';
    }
  }

  // Save last module
  if (currentModule && currentModule.number && currentModule.title) {
    currentModule.content = contentBuffer.trim();
    modules.push(currentModule as Module);
  }

  return modules;
}

function parseSkills(text: string): ESCOSkill[] {
  const skills: ESCOSkill[] = [];

  // Pattern for skills like "διοίκηση γραφείου [E]" or "διαχείριση συμβολαίων [O]"
  const skillPattern = /([^[\]\n,]+)\s*\[([EO])\]/g;
  let match;

  while ((match = skillPattern.exec(text)) !== null) {
    const name = match[1].trim();
    const type = match[2] === 'E' ? 'essential' : 'optional';

    if (name && name.length > 2) {
      skills.push({
        code: generateSkillCode(name),
        name,
        type,
      });
    }
  }

  // If no bracketed skills found, try comma-separated list
  if (skills.length === 0) {
    const parts = text.split(/[,\n]/);
    for (const part of parts) {
      const name = part.trim();
      if (name && name.length > 3 && !name.match(/^\d/) && !name.includes('ESCO')) {
        skills.push({
          code: generateSkillCode(name),
          name,
          type: 'essential',
        });
      }
    }
  }

  return skills;
}

function generateSkillCode(name: string): string {
  // Generate a simple code from the skill name
  return name
    .toLowerCase()
    .replace(/[^\w\sα-ωά-ώ]/g, '')
    .split(/\s+/)
    .map(word => word.substring(0, 3))
    .join('-')
    .substring(0, 20);
}

export function formatModuleForGeneration(module: Module): string {
  return `
## Ενότητα ${module.number}: ${module.title}

**Διάρκεια:** ${module.hours} ώρες

**Περιεχόμενο:**
${module.content}

**Δραστηριότητες:**
${module.activities}

**Δεξιότητες ESCO:**
${module.skills.map(s => `- ${s.name} (${s.type === 'essential' ? 'Απαραίτητη' : 'Προαιρετική'})`).join('\n')}
`.trim();
}
